Network/Socket/Types.hsc

```diff
@@ -910,6 +914,9 @@ withSockAddr addr f = do
     let sz = sizeOfSockAddr addr
     allocaBytes sz $ \p -> pokeSockAddr p addr >> f (castPtr p) sz
 
+unixPathMax :: Int
+unixPathMax = 108
+
```

Comment:
Note that on NetBSD, FreeBSD and Darwin the size of `sun_path` in `sockaddr_un` is 104 bytes. However, allocating larger `sockaddr` buffers (than `sizeof(sockaddr_un)`) allows use of longer (NUL-terminated) pathnames.  So we can in fact bind and connect to sockets with names that are 108 (or more) bytes in length on these systems, provided the `sockaddr` buffer is long enough.  The only issue is that not all clients will be able to interoperate with sockets whose names that exceed the default `sizeof(sun_path)` of `sockaddr_un`, but that's the user's problem if he/she chooses to create such sockets.

If there are still Linux systems that fail to NUL-terminate returned socket addresses, we should however promise **to** the OS one less byte than we allocated and zeroed:
```
withNewSocketAddress f = allocaBytes sockaddrStorageLen $ \ptr -> do
     zeroMemory ptr $ fromIntegral sockaddrStorageLen
     f ptr (sockaddrStorageLen-1)
```

That said, note that this pull request uses `sockaddrStorageLen = 128` only to allocate storage for socket addresses we get back  **from** the OS in `withNewSocketAddress` but still leaves
```
sizeOfSockAddr SockAddrUnix{}  = #const sizeof(struct sockaddr_un)
```
as the size used in `withSocketAddress` to allocate socket addresses for passing **to** the OS.  Therefore, on BSD and Darwin pathnames of 108 bytes will overflow the allocated buffer when binding paths of 108 bytes.

Since we need to support initialization of the same size sockets as we're willing to receive (and in particular path of up 108 bytes on BSD and the like, where this requires socket paths longer than fit in the stock `sockaddr_un`) we need to also allocate space for `sockaddrStorageLen` bytes when initializing `AF_UNIX` sockaddrs with a user provided `path`.
Since we'll refuse to actually write more than 108 bytes of `path` in `pokeSockAddr`, the result will always be NUL-terminated. 
```
sizeOfSockAddr SockAddrUnix{}  = #const sizeof(struct sockaddr_un)
```
to
```
sizeOfSockAddr SockAddrUnix{path}  = sockaddrStorageLen
```

Reply:
Thank you for this investigation. I agree with your approach. So, I pushed the fix.